clang/test/Modules/friend-definition-2.cpp

```diff
@@ -1,32 +1,53 @@
-// RUN: %clang_cc1 -std=c++14 -fmodules %s -verify
-// RUN: %clang_cc1 -std=c++14 -fmodules %s -verify -triple i686-windows
-// expected-no-diagnostics
-#pragma clang module build A
-module A {}
-#pragma clang module contents
-#pragma clang module begin A
+// RUN: split-file %s %t
+
+// RUN: %clang_cc1 -std=c++14 -x c++ -fmodules -fmodule-name=A -emit-module %t/a.modulemap -o %t/a.pcm
+// RUN: %clang_cc1 -std=c++14 -x c++ -fmodules -fmodule-name=B -emit-module %t/b.modulemap -o %t/b.pcm
+// RUN: %clang_cc1 -std=c++14 -x c++ -fmodules -fmodule-map-file=%t/a.modulemap -fmodule-map-file=%t/b.modulemap \
+// RUN:   -fmodule-file=%t/a.pcm -fmodule-file=%t/b.pcm \
+// RUN:   %t/use.cc -verify
+
+// RUN: rm -f %t/*.pcm
+
+// RUN: %clang_cc1 -std=c++14 -x c++ -fmodules -fmodule-name=A -emit-module %t/a.modulemap -o %t/a.pcm -triple i686-windows
+// RUN: %clang_cc1 -std=c++14 -x c++ -fmodules -fmodule-name=B -emit-module %t/b.modulemap -o %t/b.pcm -triple i686-windows
+// RUN: %clang_cc1 -std=c++14 -x c++ -fmodules -fmodule-map-file=%t/a.modulemap -fmodule-map-file=%t/b.modulemap \
+// RUN:   -fmodule-file=%t/a.pcm -fmodule-file=%t/b.pcm \
+// RUN:   %t/use.cc -verify -triple i686-windows
+
+//--- a.modulemap
+module A {
+  header "a.h"
+}
+
+//--- a.h
+#ifndef A_H
+#define A_H
+template<typename T> struct ct { friend auto operator-(ct, ct) { struct X {}; return X(); } void x(); };
+#endif
+
+//--- b.modulemap
+module B {
+  header "b.h"
+}
+
+//--- b.h
+#ifndef B_H
+#define B_H
 template<typename T> struct ct { friend auto operator-(ct, ct) { struct X {}; return X(); } void x(); };
```

Comment:
I think it is better to replace it with `#include "a.h"` and it will have exactly the same issue with  #111992

Reply:
Because header modules rewrite #include as imports in that case [1], this would actually change the semantics of the code.

I.e. later in `use.cc`, we would be importing from the module `A` rather than merging from the module `A` and `B`.

[1] assuming we would pass `-fmodule-map-file` and `-fmodule-file` and create a dependency chain. It's also possible to mix the same header as modular and non-modular across multiple compilations, but it quickly becomes too confusing for anyone to follow.

Reply:
I think in this particular case it doesn't change semantic because you don't pass `a.modulemap` when compiling `b.cpm` so it is textual inclusion. To be 100% compatible you can create a new header without modules map and textually include it in both headers (just having code duplication is not a good idea and also doesn't match what happens in reality) but I'm OK with keeping it as-is.

Reply:
Yeah, that's correct, I somehow thought you meant also passing the module map. Using the same header as modular and non-modular in the same test is quite hard to follow, so I didn't want to go for it. Introducing a new textual header totally works.

Sharing the same header would've been an option, but since the file is small, hopefully it doesn't matter much.